### ISSUE TITLE
Type inference error for heterogeneous array

### ISSUE THREAD

**Issue by petersolopov** [microsoft/TypeScript#33664]
**Search Terms:** Generic, Heterogeneous Array

**Code**

```ts
function shuffleArray<I>(array: I[]): I[] {
  return (
    array
      .map(item => [Math.random(), item])
      .sort((a, b) => a[0] - b[0])
      .map(item => item[1])
  );
}
```

**Expected behavior:** no error

**Actual behavior:** error
```
Type '(number | I)[]' is not assignable to type 'I[]'.
  Type 'number | I' is not assignable to type 'I'.
    'number | I' is assignable to the constraint of type 'I', but 'I' could be instantiated with a different subtype of constraint '{}'.
      Type 'number' is not assignable to type 'I'.
        'number' is assignable to the constraint of type 'I', but 'I' could be instantiated with a different subtype of constraint '{}'.
```

**Playground Link:** http://www.typescriptlang.org/play/#code/GYVwdgxgLglg9mABAZwBYmMANgUwIIBOBAhgJ4A8AkgHwAUxRZAXIpQNoC6AlC+x4gG8AUIkQEcUEASS0RoxAxKk58xADoAtsQAOtGFBwbEAXmqI2AWWJRUakmAAmcDbS4AaRPsPcV8tcjgCKFp6DwAjLhMzYjYABn4AWkQwuJ9VUU0dPQMjU08ctgBGNMQuAG4hAF8gA

**P.S** Flow works [fine](https://flow.org/try/#0GYVwdgxgLglg9mABAZwBYmMANgUwIIBOBAhgJ4A8AkgHwAUxRZAXIpQNoC6AlC+x4gG8AUIkQEcUEASS0RoxAxKk58xADoAtsQAOtGFBwbEAXmqI2AWWJRUakmAAmcDbS4AaRPsPcV8tcjgCKFp6DwAjLhMzYjYABn4AWkQwuJ9VUU0dPQMjU08ctgBGNMQuAG4hAF8gA)

**Comment by jablko** [comment]
Does adding [`as const`](https://devblogs.microsoft.com/typescript/announcing-typescript-3-4/#const-assertions) help? e.g.
```TypeScript
.map(item => [Math.random(), item] as const)
```

**Comment by petersolopov** [comment]
@jablko Yes, thanks! But I still think It should work without this feature

**Comment by MartinJohns** [comment]
It does, if you're explicit with the types:

    .map<[number, I]>(item => [Math.random(), item])

By default array-literals are inferred as arrays, so you return an array that has either a `number` or a `I`. Changing this to automatically infer as tuples would very likely be a breaking change. And we already have a way to tell the compiler to infer it as a tuple: `as const`.
